File: src/algorithms/heaps/construction/build-max-heap/build-max-heap.test.ts
import { describe, it, expect } from "vitest";
import { buildMaxHeap } from "./sources/build-max-heap.ts?fn";

/** Verify max-heap property: every parent ≥ both children. */
function isMaxHeap(array: number[]): boolean {
  const size = array.length;
  for (let parentIdx = 0; parentIdx < Math.floor(size / 2); parentIdx++) {
    const leftIdx = 2 * parentIdx + 1;
    const rightIdx = 2 * parentIdx + 2;
    if (leftIdx < size && array[parentIdx]! < array[leftIdx]!) return false;
    if (rightIdx < size && array[parentIdx]! < array[rightIdx]!) return false;
  }
  return true;
}

describe("buildMaxHeap", () => {
  it("produces a valid max-heap for the default 9-element input", () => {
    const result = buildMaxHeap([9, 5, 7, 1, 3, 8, 2, 6, 4]) as number[];
    expect(isMaxHeap(result)).toBe(true);
  });

  it("root is the maximum element", () => {
    const result = buildMaxHeap([9, 5, 7, 1, 3, 8, 2, 6, 4]) as number[];
    expect(result[0]).toBe(9);
  });

  it("preserves all elements (same multiset)", () => {
    const input = [9, 5, 7, 1, 3, 8, 2, 6, 4];
    const result = buildMaxHeap(input) as number[];
    expect([...result].sort((a, b) => a - b)).toEqual([...input].sort((a, b) => a - b));
  });

  it("handles an already valid max-heap", () => {
    const input = [9, 7, 8, 5, 6, 3, 4];
    const result = buildMaxHeap(input) as number[];
    expect(isMaxHeap(result)).toBe(true);
    expect(result[0]).toBe(9);
  });

  it("handles a sorted ascending (min-heap) array", () => {
    const result = buildMaxHeap([1, 2, 3, 4, 5, 6, 7]) as number[];
    expect(isMaxHeap(result)).toBe(true);
    expect(result[0]).toBe(7);
  });

  it("handles a single element", () => {
    const result = buildMaxHeap([42]) as number[];
    expect(result).toEqual([42]);
  });

  it("handles two elements", () => {
    const result = buildMaxHeap([2, 5]) as number[];
    expect(result[0]).toBe(5);
    expect(isMaxHeap(result)).toBe(true);
  });

  it("does not mutate the original array", () => {
    const input = [9, 5, 7, 1, 3];
    buildMaxHeap(input);
    expect(input).toEqual([9, 5, 7, 1, 3]);
  });
});

File: src/algorithms/heaps/construction/build-max-heap/sources/build-max-heap_test.go
package heaps

import "testing"

func isMaxHeapBMH(array []int) bool {
	size := len(array)
	for parentIdx := 0; parentIdx < size/2; parentIdx++ {
		leftIdx := 2*parentIdx + 1
		rightIdx := 2*parentIdx + 2
		if leftIdx < size && array[parentIdx] < array[leftIdx] {
			return false
		}
		if rightIdx < size && array[parentIdx] < array[rightIdx] {
			return false
		}
	}
	return true
}

func TestBuildMaxHeapValid(t *testing.T) {
	result := buildMaxHeap([]int{9, 5, 7, 1, 3, 8, 2, 6, 4})
	if !isMaxHeapBMH(result) {
		t.Errorf("Result is not a valid max-heap: %v", result)
	}
}

func TestBuildMaxHeapRootIsMax(t *testing.T) {
	result := buildMaxHeap([]int{9, 5, 7, 1, 3, 8, 2, 6, 4})
	if result[0] != 9 {
		t.Errorf("Expected root=9, got %d", result[0])
	}
}

func TestBuildMaxHeapAlreadyValid(t *testing.T) {
	result := buildMaxHeap([]int{9, 7, 8, 5, 6, 3, 4})
	if !isMaxHeapBMH(result) || result[0] != 9 {
		t.Errorf("Expected valid max-heap with root=9, got %v", result)
	}
}

func TestBuildMaxHeapSortedAscending(t *testing.T) {
	result := buildMaxHeap([]int{1, 2, 3, 4, 5, 6, 7})
	if !isMaxHeapBMH(result) || result[0] != 7 {
		t.Errorf("Expected valid max-heap with root=7, got %v", result)
	}
}

func TestBuildMaxHeapSingle(t *testing.T) {
	result := buildMaxHeap([]int{42})
	if len(result) != 1 || result[0] != 42 {
		t.Errorf("Expected [42], got %v", result)
	}
}

func TestBuildMaxHeapTwoElements(t *testing.T) {
	result := buildMaxHeap([]int{2, 5})
	if result[0] != 5 || !isMaxHeapBMH(result) {
		t.Errorf("Expected max-heap with root=5, got %v", result)
	}
}
